notes/fomega/evaluation/builtin-evaluation.md

```diff
@@ -0,0 +1,285 @@
+CEK-value is instantiated or applied to an argument the argument is
+added to the appropriate list and the count is decremented.  When the
+count reaches zero the builtin name and the arguments can be passed
+directly to the constant application machinery.
+
+(This technique is slightly fragile because we only count the total
+number of type and term arguments, and don't check that type and term
+arguments are correctly interleaved.  However, (a) these _will_ be
+typed by the typechecker, (b) we'll eventually be moving to an untyped
+evaluator where this is less important, and (c) if we really wanted to
+be more robust we could keep separate lists of type and term arities,
+or (better, but at a slight cost) provide a list of `TypeArg |
+TermArg` values, removing the first element every time we encounter an
+instantiation or application, and checking that it is of the correct kind.)
+
+This process is clearly much more efficient than that in the original
+CEK machine, and in fact enabled the alternative CEK machine with
+unsaturated builtins to become competitive with the one with saturated
+builtins.
+
+
+### The CK machine
+
+Unfortunately we can't use the above technique with the CK machine:
+that works directly with values as described in the Plutus Core
+specification, and there's nowhere to store the argument information.
+I've considered a number of ways to fix this.
+
+ * I initially thought we could store the required information
+   in a frame.  This is how the CK machine worked with saturated
+   builtins, but we can't use the same technique with unsaturated
+   builtins.  In the unsaturated case a partially applied builtin
+   is a perfectly legitimate value and could be returned by a function
+   or supplied as an argument to another function.  This just doesn't
+   fit in with the way that stack frames are used.
+
+ * We could modify the notion of value used in the CK machine. This
+   might be sensible.  The CK machine and the CEK machine both have two
+   main phases: "compute" and "return", the former acting on terms and
+   the latter acting on values.  In the CK machine, we've been using
+   values of the type described in the Plutus Core specification,
+   which are terms with a particular syntactic form.  However, I think
+   there's no reason why type of the arguments consumed by the return
+   phase have to bear any relation to the term type consumed by the
+   compute phase, so this might be the way to go.
+
+   UPDATE: I've now done this and it seems to perform OK: it may be 1%
+   or so slower than the intiial solution described below.  I added a
+   `CkValue` type that was more or less isomorphic to the usual values
+   except for builtins, where the value includes a count of arguments
+   and lists of the type and term arguments it's received so far.
+   Introducing the new type also required quite a lot of typeclass
+   stuff to make it acceptable to the CAM.  See
```

Comment:
Yeah.  In the earlier version of the CK machine it didn't need these because it was using Terms, which already have suitable instances defined somewhere.  Of course, when I tried to use CkValues instead I started to get terrifying type errors. Looking at it again, there's less stuff required to fix that than I thought.